### DISCUSSION TITLE
Externalize Propagation of Elements to Queue

### DISCUSSION THREAD

**Discussion by deeekay** [craftcms/cms#13977] [Ideas] [answered]
Hi there

We are currently building a craft website with a lot of sites. In combination with matrix and neo blocks, saving entries (with the propagation to all sites) can become very slow. So we were wondering if there is (or could be in the future) a possibility to externalize the propagation of content from elements to other sites within a queue to improve response times?

Thanks for your response in advance and best regards,
Daniel

**Comment by brandonkelly** [comment]
I’ve considered this. It would drastically complicate the logic, and make the authoring flow a bit more complex—we’d need to figure out the right approach for handling edits to elements/sites which are pending propagation—but maybe worth it for more complex content models.

Craft 5 has a couple significant Matrix performance improvements, which may tip the scale and make this less needed.

**Reply by deeekay** [reply]
Thanks for the prompt reply @brandonkelly 🙏🏽

Do you see any other way of improving this on our side or optimize in order to reduce saving times when propagating to other sites?

Thanks a lot and best regards,
Daniel

**Reply by brandonkelly** [reply, marked as answer]
The best thing you could do is find ways to simplify the content model, with less Matrix/Neo/Super Table nesting.

**Reply by deeekay** [reply]
@brandonkelly I was thinking if it could be a solution to handle the saving as follows:

1. User triggers "Save Entry" action
2. Entry (of the current site) is saved (synchronously)
3. A background job is forked that propagates the content for all affected sites (asynchronously)
4. In the meantime, the original entry reloads fast and shows a message that the content is being propagated (optionally with a nice progress bar 😊 ) making all affected content read-only (for the time the background job is running)
5. As soon as the propagation is completed, the message disappears and the content is rendered editable again.

This would result in a fast response time and allow the user to go about their work.

Could that work?

Best regards,
Daniel

**Reply by brandonkelly** [reply]
That’s what I assumed you meant by the original post.

It just gets really tricky to maintain data integrity once you start splitting the save up into multiple steps, which could potentially overlap other conflicting save operations. So we’d either need to lock the entry to avoid conflicting edits (which we already do when saving, but it would be significantly more noticeable when it spans into a queue job), or add a complex change tracking system.

**Reply by deeekay** [reply]
@brandonkelly yeah, locking entries for the time of the save action (and the subsequent asynchronous actions) was what I was trying to propose 😊 A complex change tracking would - most likely - be an overkill, at least for the challenge we are facing at the moment. Is there a way in which we could contribute to solving this issue (conceptually and/or technically)?

Best regards,
Daniel
